File: CS/DataProvider.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Xml;
using System.Web;
using System.Web.SessionState;

public static class DataProvider {
    private const string SessionKey = "CountriesSampleDataSource";
    private const string XmlFileVirtPath = "~/App_Data/Countries.xml";

    private static HttpSessionState Session {
        get { return HttpContext.Current.Session; }
    }
    private static string CountiesXmlFilePhysPath {
        get { return HttpContext.Current.Request.MapPath(XmlFileVirtPath); }
    }

    public static IList<string> GetCountries() {
        IList<string> countries = (IList<string>)Session[SessionKey];
        if(countries == null) {
            countries = new ReadOnlyCollection<string>(LoadCountries());
            Session[SessionKey] = countries;
        }
        return countries;
    }

    private static List<string> LoadCountries() {
        List<string> result = new List<string>();
        XmlDocument doc = new XmlDocument();
        doc.Load(CountiesXmlFilePhysPath);
        if(doc.ChildNodes.Count <= 1)
            throw new System.IO.InvalidDataException("Unable to read data file.");
        XmlNode countriesRootNode = doc.ChildNodes[1];
        result.Capacity = countriesRootNode.ChildNodes.Count;
        foreach(XmlNode countryNode in countriesRootNode.ChildNodes)
            result.Add(countryNode.Attributes["name"].Value);
        return result;
    }
}

File: VB/DataProvider.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports System.Xml
Imports System.Web
Imports System.Web.SessionState

Public NotInheritable Class DataProvider
	Private Const SessionKey As String = "CountriesSampleDataSource"
	Private Const XmlFileVirtPath As String = "~/App_Data/Countries.xml"

	Private Sub New()
	End Sub
	Private Shared ReadOnly Property Session() As HttpSessionState
		Get
			Return HttpContext.Current.Session
		End Get
	End Property
	Private Shared ReadOnly Property CountiesXmlFilePhysPath() As String
		Get
			Return HttpContext.Current.Request.MapPath(XmlFileVirtPath)
		End Get
	End Property

	Public Shared Function GetCountries() As IList(Of String)
		Dim countries As IList(Of String) = CType(Session(SessionKey), IList(Of String))
		If countries Is Nothing Then
			countries = New ReadOnlyCollection(Of String)(LoadCountries())
			Session(SessionKey) = countries
		End If
		Return countries
	End Function

	Private Shared Function LoadCountries() As List(Of String)
		Dim result As List(Of String) = New List(Of String)()
		Dim doc As New XmlDocument()
		doc.Load(CountiesXmlFilePhysPath)
		If doc.ChildNodes.Count <= 1 Then
			Throw New System.IO.InvalidDataException("Unable to read data file.")
		End If
		Dim countriesRootNode As XmlNode = doc.ChildNodes(1)
		result.Capacity = countriesRootNode.ChildNodes.Count
		For Each countryNode As XmlNode In countriesRootNode.ChildNodes
			result.Add(countryNode.Attributes("name").Value)
		Next countryNode
		Return result
	End Function
End Class